x-pack/lib/filters/geoip/database_manager.rb

```diff
@@ -30,134 +31,193 @@ module LogStash module Filters module Geoip class DatabaseManager
+        if @states[database_type].is_eula && @states[database_type].plugins.size > 0
+          logger.error("The MaxMind database hasn't been updated from last 30 days. Logstash is unable to get newer version from internet. "\
+            "According to EULA, GeoIP plugin needs to stop using MaxMind database in order to be compliant. "\
+            "Please check the network settings and allow Logstash accesses the internet to download the latest database, "\
+            "or switch to offline mode (:database => PATH_TO_YOUR_DATABASE) to use a self-managed database "\
+            "which you can download from https://dev.maxmind.com/geoip/geoip2/geolite2/ ")
+          @states[database_type].plugins.dup.each { |plugin| plugin.expire_action if plugin }
+        end
+      when days_without_update >= 25
+        if @states[database_type].is_eula && @states[database_type].plugins.size > 0
+          logger.warn("The MaxMind database hasn't been updated for last #{days_without_update} days. "\
+          "Logstash will fail the GeoIP plugin in #{30 - days_without_update} days. "\
+          "Please check the network settings and allow Logstash accesses the internet to download the latest database ")
+        end
+      else
+        logger.trace("The endpoint hasn't updated", :days_without_update => days_without_update)
+      end
+    end
+  end
+
+  # Clean up files .mmdb, .tgz which are not mentioned in metadata and not default database
+  def clean_up_database
+    protected_dirnames = (@metadata.dirnames + [CC]).uniq
+    existing_dirnames = ::Dir.children(get_data_dir_path)
+                             .select { |f| ::File.directory? ::File.join(get_data_dir_path, f) }
+
+    (existing_dirnames - protected_dirnames).each do |dirname|
+      dir_path = get_dir_path(dirname)
+      FileUtils.rm_r(dir_path)
+      logger.debug("#{dir_path} is deleted")
+    end
+  end
+
+  def trigger_download
+    if @trigger_download.false? && @trigger_download.make_true
+      execute_download_job
+
+      # check database update periodically. trigger `call` method
+      @scheduler = Rufus::Scheduler.new({:max_work_threads => 1})
+      @scheduler.every('24h', self)
     end
   end
 
+  public
+
   # scheduler callback
   def call(job, time)
     logger.debug "scheduler runs database update check"
+    execute_download_job
+  end
 
-    begin
-      if execute_download_job
-        @geoip.setup_filter(database_path)
-        clean_up_database
-      end
-    rescue DatabaseExpiryError => e
-      logger.error(e.message, :cause => e.cause, :backtrace => e.backtrace)
-      @geoip.terminate_filter
-    end
+  def database_path(database_type)
+    @states[database_type].database_path
   end
 
   def close
     @scheduler.every_jobs.each(&:unschedule) if @scheduler
   end
 
-  def database_path
-    @database_path
-  end
+  def subscribe_database_path(database_type, database_path, geoip_plugin)
+    if database_path.nil?
+      trigger_download
```

Comment:
`trigger_download()` is the entry point called by [geoip plugin](https://github.com/logstash-plugins/logstash-filter-geoip/pull/181/files#diff-5421cf35b684cca803ff336c23f52918df0809819ae2bd6d36a4a4ec8250ddf1R150)